### PR TITLE
Reduce calculation for fragments in ExecutableNormalizedOperation

### PR DESCRIPTION
**Pull request by dondonz**
After parsing and validating an incoming GraphQL operation, the engine creates a normalized version of that operation, the `ExecutableNormalizedOperation`.

For fragments, there is a little more work involved to determine the possible types. Inside `narrowDownPossibleObjects`, there is a set intersection between the current set of possible objects and the set of types from the resolved type condition.

This set intersection can be expensive for very large queries with hundreds of fragments. This PR includes one such example.

In almost every case in the example query, one of the sets being compared only has one member. For example, the resolved types set often has only 1 member, because an inline fragment on a concrete type has only one possible type. It would be equivalent, and much cheaper, to calculate via Set `contains` rather than `intersection`.

And I've added a more general tweak to always pass the smallest set first to Guava's `Sets.intersection`. The implementation is faster when the first set is smaller.

### Benchmark results
**Baseline**
```
Benchmark                                 Mode  Cnt    Score    Error  Units
NQExtraLargeBenchmark.benchMarkThroughput  thrpt    9  274.744 ± 14.909  ops/s
NQExtraLargeBenchmark.benchMarkAvgTime      avgt    9    3.658 ±  0.310  ms/op
```

**Optimised**
```
Benchmark                                 Mode  Cnt    Score    Error  Units
NQExtraLargeBenchmark.benchMarkThroughput  thrpt    9  333.346 ± 25.986  ops/s
NQExtraLargeBenchmark.benchMarkAvgTime      avgt    9    2.905 ±  0.124  ms/op
```

Note: For throughput, a higher number is better. For average time, a lower number is better.

### CODE REVIEW

File: src/main/java/graphql/normalized/ExecutableNormalizedOperationFactory.java [hunk 1]
```diff
@@ -453,6 +453,15 @@ private Set<GraphQLObjectType> narrowDownPossibleObjects(Set<GraphQLObjectType>
...
             return resolvedTypeCondition;
         }
+
+        // Intersection calculation is expensive when either set is large.
+        // If a set only has one member, it is equivalent and cheaper to calculate via contains.
+        if (resolvedTypeCondition.size() == 1 && currentOnes.contains(resolvedTypeCondition.iterator().next())) {
+            return resolvedTypeCondition;
+        } else if (currentOnes.size() == 1 && resolvedTypeCondition.contains(currentOnes.iterator().next())) {
+            return currentOnes;
+        }
```
Comment (bbakerman) [inline src/main/java/graphql/normalized/ExecutableNormalizedOperationFactory.java:463]: Java shits me on how access to a first member is that god awful `collection.iterator().next()`

But here we are!

Comment (bbakerman) [reply src/main/java/graphql/normalized/ExecutableNormalizedOperationFactory.java:463]: Perhaps create a utitlity method in `graphql.util.FpKit`

`FpKit. intersection(` that does this optimisation for same type sets

**Review by bbakerman — APPROVED** [review]

File: src/main/java/graphql/normalized/ExecutableNormalizedOperationFactory.java (hunk 1 above)
Comment (dondonz) [reply src/main/java/graphql/normalized/ExecutableNormalizedOperationFactory.java:463]: That's a good idea I'll stick it in `FpKit` instead

File: src/main/java/graphql/util/FpKit.java [hunk 2]
```diff
@@ -330,4 +330,27 @@ public static <T> Supplier<T> interThreadMemoize(Supplier<T> delegate) {
...
+        } else if (set2.size() == 1 && set1.contains(set2.iterator().next())) {
+            return set2;
+        }
+
+        // Guava's Sets.intersection is faster when the smaller set is passed as the first argument.
+        if (set1.size() < set2.size()) {
+            return Sets.intersection(set1, set2);
+        }
+        return Sets.intersection(set2, set1);
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/util/FpKit.java:354]: There should be a test for this.  I know its simple.  So lock it in with a test

**Review by bbakerman — CHANGES_REQUESTED** [review]
Add tests for the new intersection method

**Pull request merged**